students/guillaume/session07/html_render.py

```diff
@@ -0,0 +1,129 @@
+
+
+class Element():
+    tag = ''
+    indent = ''
+
+    def __init__(self, content=None, **kwargs):
+        self.kwargs = kwargs
+        if content is None:
+            self.content = []
+        else:
+            self.content = [content]
+
+    def append(self, content=None, **kwargs):
+        self.kwargs.update(kwargs)
+        if content is not None:
+            self.content.append(content)
+
+    def render(self, file_obj, ind=""):
+
+        # add attribute to the tag
+        att = ', '.join(['{}="{}"'.format(arg, self.kwargs[arg])
+                        for arg in self.kwargs])
+        if att != "":
+            att = ' {}'.format(att)
+        file_obj.write('{}<{}{}>\n'.format(ind, self.tag, att))
+
+        for each in self.content:
+            if hasattr(each, 'render'):
+                each.render(file_obj, ind + each.indent)
+            else:
+                file_obj.write('{}{}\n'.format(ind + 4 * ' ', each))
+        file_obj.write('{}</{}>\n'.format(ind, self.tag))
+
+
+class OneLineTag(Element):
+
+    def render(self, file_obj, ind=""):
+        att = ', '.join(['{}="{}"'.format(arg, self.kwargs[arg])
+                        for arg in self.kwargs])
+        if att != "":
+            att = ' {}'.format(att)
+        output_ls = [ind, self.tag, att, ', '.join(self.content), self.tag]
+        output = '{}<{}{}> {} </{}>\n '.format(*output_ls)
+        file_obj.write(output)
+
+
+class H(OneLineTag):
+    tag = 'h'
+    indent = 4 * ' '
+
+    def __init__(self, level, content=None):
+        OneLineTag.__init__(self, content)
+        self.tag = '{}{}'.format(self.tag, str(level))
+
+
+class SelfClosingTag(Element):
+
+    def render(self, file_obj, ind=""):
+        att = ', '.join(['{}="{}"'.format(arg, self.kwargs[arg])
```

Comment:
This looks a lot like the code in the Element render method -- maybe factor that out so as not to repeat code.